solr/core/src/java/org/apache/solr/update/DocumentBuilder.java

```diff
@@ -258,6 +246,48 @@ public static Document toDocument(SolrInputDocument doc, IndexSchema schema, boo
     return out;
   }
 
+  private static boolean addOriginalField( Object originalFieldValue, SchemaField sfield, boolean forInPlaceUpdate, Document out, Set<String> usedFields) {
+    addField(out, sfield, originalFieldValue, forInPlaceUpdate);
+    // record the field as having a value
+    usedFields.add(sfield.getName());
+    return true;
+  }
+
+  private static boolean addCopyFields(Object originalFieldValue, FieldType originalFieldType, List<CopyField> copyFields, boolean forInPlaceUpdate, String uniqueKeyFieldName, Document out, Set<String> usedFields) {
+    boolean used = false;
+    for (CopyField cf : copyFields) {
+      SchemaField destinationField = cf.getDestination();
+
+      final boolean destHasValues = usedFields.contains(destinationField.getName());
+
+      // Dense Vector Fields can only be copied to same field type
+      if (originalFieldType instanceof DenseVectorField && !(destinationField.getType() instanceof DenseVectorField)) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+                "The copy field destination must be a DenseVectorField: " +
+                        destinationField.getName());
+      }
```

Comment:
Is there also a requirement for dimension alignment i.e. `dst` must not be higher dimension than `src` and/or they should be equal? Maybe that's already covered elsewhere, I haven't looked.

Reply:
Yes, you'll get an exception telling that, just added the test!